### DISCUSSION TITLE
Completer function seems to breaks tab and other special keys

### DISCUSSION THREAD

**Discussion by yggdr** [xonsh/xonsh#5516] [Q&A]
I'm currently trying to fix the [carapace](https://github.com/carapace-sh/carapace-bin) xonsh integration. I stumbled upon a strange behaviour that I cannot seem to pin to a bug in the carapace-generated xonsh code, so I'm wondering if this might be an issue in xonsh itself.

The strange behaviour:
Completion for which carapace has no support configured works unchanged.
When completing a command carapace supports, characters special behaviour breaks. When I input `ls -<Tab>` to start the completion, nothing happens (although I have also seen it at least displaying the completion options carapace offered as expected at times).
Any further pressing of special characters like the Tab of Left or Right Arrow key now does not trigger their associated special action anymore (like cycling though the completions, or going left or right by one character), instead, it inserts them literally (like a literal `	`, arrow keys become their escape codes `^[[C` etc.)

Carapace works by installing a lazy completer at the start of the completers list. When a completion is triggered, this function checks if it can provide a completion for this situation, and if so, generates a completer function, installs it at the start of the completers list, and returns it.
The part where it installs the new function at the start of the completers list during the triggered completion has lead to its very own mutating-an-OrderedDict-while-iterating-over-it issue, so I thought that might be the root cause.
So I took the generated completion function for the `ls` command, installed it manually in a freshly installed 0.17.0.dev1 xonsh run with `--no-rc`, and still see the same behaviour.

`carapace ls xonsh` generates the following completer function:
```xsh
from xonsh.completers.tools import contextual_command_completer, RichCompletion
from xonsh.completers.completer import add_one_completer
import json

@contextual_command_completer
def _ls_completer(context):
    """carapace completer for ls"""
    if context.completing_command('ls'):
        def fix_prefix(s):
            """quick fix for partially quoted prefix completion ('prefix',<TAB>)"""
            return s.translate(str.maketrans('', '', '\'"'))

        output = __xonsh__.subproc_captured_stdout(['carapace', 'ls', 'xonsh', *[a.value for a in context.args], fix_prefix(context.prefix)])
        try:
            result = {RichCompletion(c["Value"], display=c["Display"], description=c["Description"], prefix_len=len(context.raw_prefix), append_closing_quote=False, style=c["Style"]) for c in json.loads(output)}
        except:
            result = set()
        if len(result) == 0:
            result = {RichCompletion(context.prefix, display=context.prefix, description='', prefix_len=len(context.raw_prefix), append_closing_quote=False)}
        return result

add_one_completer('ls', _ls_completer, 'start')
```
Install by `evalx(compilex($(carapace ls xonsh)))`, then try the completion.

As you can see, the generated completer functions essentially calls `carapace ls xonsh ls -`, which generates the completions as a JSON-string. That commands output looks good to me, and setting `$XONSH_TRACE_COMPLETIONS = True` also shows nothing unexpected/different to other, non-carapace completers:

carapace completer
```xsh
$ ls -<Tab>
TRACE COMPLETIONS: Getting completions with context:
CompletionContext(command=CommandContext(args=(CommandArg(value='ls', opening_quote='', closing_quote='', is_io_redir=False),), arg_index=1, prefix='-', suffix='', opening_quote='', closing_quote='', is_after_closing_quote=False, subcmd_opening=''), python=PythonContext('ls -', 4, is_sub_expression=False))
TRACE COMPLETIONS: Got 82 results from exclusive completer 'ls':
[(RichCompletion('--ignore-backups ', prefix_len=1, display='--ignore-backups', description='do not list implied entries ending with ~', style='bg:default fg:default', append_closing_quote=False),
  1),
 (RichCompletion('-D ', prefix_len=1, display='-D', description="generate output designed for Emacs' dired mode", style='bg:default fg:default', append_closing_quote=False),
  1),
[…]
(RichCompletion('--show-control-chars ', prefix_len=1, display='--show-control-chars', description='show nongraphic characters as-is (the default,', style='bg:default fg:default', append_closing_quote=False),
  1)]
```

Non-carapace completer
```xsh
$ less -<Tab>
less -
TRACE COMPLETIONS: Getting completions with context:
CompletionContext(command=CommandContext(args=(CommandArg(value='less', opening_quote='', closing_quote='', is_io_redir=False),), arg_index=1, prefix='-', suffix='', opening_quote='', closing_quote='', is_after_closing_quote=False, subcmd_opening=''), python=PythonContext('less -', 6, is_sub_expression=False))
TRACE COMPLETIONS: Got 83 results from exclusive completer 'man':
[(RichCompletion('--help', prefix_len=1, display='--help', description='This option displays a summary of the commands accepted by less\n(the same as the h command).  (Depending on how your shell\ninterprets the question mark, it may be necessary to quote the\nquestion mark, thus: "-\\?".)'),
  1),
[…]
 (RichCompletion('--wordwrap', prefix_len=1, display='--wordwrap', description='When the -S option is not in use, wrap each line at a space or\ntab if possible, so that a word is not split between two lines.\nThe default is to wrap at any character.'),
  1)]
```

As usual, any help would be appreciated :)

**Comment by anki-code** [comment]
@jnoortheen could you please take a look into the completer implementation. After massive refactoring before I'm afraid to give wrong advice.

**Reply by anki-code** [reply]
@yggdr to achieve syntax highlighting just add xsh e.g.:

<img width="82" alt="image" src="https://github.com/xonsh/xonsh/assets/1708680/af0cf5e9-d552-44ca-9b73-4aa82e62784c">

**Comment by jnoortheen** [comment]
@yggdr if the generated code for each of the command is not that different, you could try something like this for the integration 

```py
from functools import lru_cache

from xonsh.completers.completer import add_one_completer
from xonsh.completers.tools import contextual_command_completer, sub_proc_get_output


@lru_cache
def get_supported_cmds():
    output, _ = sub_proc_get_output("carapace", "--list")
    if not output:
        return []
    return {line.split()[0] for line in output.decode().splitlines()}


@contextual_command_completer
def _carapace_lazy(context):
    """carapace completer for aws"""
    supported_cmds = get_supported_cmds()

    if context.command not in supported_cmds:
        return

    from json import loads

    from xonsh.completers.tools import RichCompletion

    def fix_prefix(s):
        """quick fix for partially quoted prefix completion ('prefix',<TAB>)"""
        return s.translate(str.maketrans('', '', '\'"'))

    output, _ = sub_proc_get_output(
        'carapace', context.command, 'xonsh', *[a.value for a in context.args], fix_prefix(context.prefix))
    if not output:
        return

    for c in loads(output):
        yield RichCompletion(
            c["Value"],
            display=c["Display"],
            description=c["Description"],
            prefix_len=len(context.raw_prefix),
            append_closing_quote=False,
            style=c["Style"],
        )


add_one_completer('carapace_lazy', _carapace_lazy, '>xompleter')
```

**Reply by yggdr** [reply]
Thank you, this code seems to work quite well. What I don't understand is _why_ this solution of yours works, and `carapace`'s original doesn't. What's the difference?

**Reply by jnoortheen** [reply]
It just simplifies loading completions from carapace directly. I think this can be the default script from carapace. If possible, I will create an issue there.

**Reply by yggdr** [reply]
Yes, that part I understand, and had planned to do something like it myself once this issue of breaking tab-completion was resolved. Thing is, I don't understand why your code unbreaks tab-completion. To me it seems it does more or less the same as the original.
